### PR TITLE
Add NOTIFICATIONS_SYSTEM_LEVEL_TRUMP

### PR DESCRIPTION
**Pull request by kiblik**
Revert a small part from #9227.

Full explanation in https://github.com/DefectDojo/django-DefectDojo/pull/8824#issuecomment-1986098912

### CODE REVIEW

**Review by mtesauro — APPROVED** [review]
Approved

**Review by grendel513 — APPROVED** [review]

**Review by Maffooch — CHANGES_REQUESTED** [review]
Dropped a comment in the other PR since that is where the bulk of the context around this PR is: https://github.com/DefectDojo/django-DefectDojo/pull/8824#issuecomment-1988747569

Maybe we could add an additional check for the `event` type to determine if the system level notifications should trump the user level. Something like this (I could get suggestions to work here):
```
if event in ["user_mentioned", "review_requested"]:
    # merge the system level notifications with the personal level
    # this allows for system to trump the personal
    merged_notifications = Notifications.merge_notifications_list([system_notifications, recipient_notifications])
    merged_notifications.user = recipient_notifications.user
    logger.debug('Sent notification to %s', merged_notifications.user)
    process_notifications(event, merged_notifications, **kwargs)
else:
    # Do not trump user preferences and send notifications as usual
    logger.debug('Sent notification to %s', recipient_notifications.user)
    process_notifications(event, recipient_notifications, **kwargs)
```

**Comment by kiblik** [comment]
> Dropped a comment in the other PR since that is where the bulk of the context around this PR is: [#8824 (comment)](https://github.com/DefectDojo/django-DefectDojo/pull/8824#issuecomment-1988747569)
> 
> Maybe we could add an additional check for the `event` type to determine if the system level notifications should trump the user level. Something like this (I could get suggestions to work here):
> 
> ```
> if event in ["user_mentioned", "review_requested"]:
>     # merge the system level notifications with the personal level
>     # this allows for system to trump the personal
>     merged_notifications = Notifications.merge_notifications_list([system_notifications, recipient_notifications])
>     merged_notifications.user = recipient_notifications.user
>     logger.debug('Sent notification to %s', merged_notifications.user)
>     process_notifications(event, merged_notifications, **kwargs)
> else:
>     # Do not trump user preferences and send notifications as usual
>     logger.debug('Sent notification to %s', recipient_notifications.user)
>     process_notifications(event, recipient_notifications, **kwargs)
> ```

Thank you @Maffooch for your reaction.
Purely technically, yes, this sounds like an alternative that would work.
However,
- This creates an anti-pattern in the systematic handling of notification
  - I suppose it puts evaluation (of how and if a notification should be processed) logic in the place where it should not be.
- It would create undo documented behavior solving one (sorry, two) specific use-case(s).

**Comment by grendel513** [comment]
I believe @Maffooch thought process and approach seems the best here

**Comment by mtesauro** [comment]
@grendel513 Thanks for chiming in - I've been meaning to comment on this for far too long.

Basically, I agree with the idea of checking the event from @Maffooch - I'm thinking that if I were a user who opted out of a notification, I'd be irritated if I got one anyway - excluding mentions or specific requests for me.

Matt's 2 cents

**Comment by kiblik** [comment]
Ok, I will change the implementation to what @Maffooch offered.
But I will add an option to customize a list of events via EnvVars (loaded in Settings) with default `["user_mentioned", "review_requested"]` - to allow admins to easily change this behavior.

**Comment by kiblik** [comment]
Hi @mtesauro / @devGregA / @Maffooch / @cneill / @blakeaowens. I would like to ask for a review.
It is an enabler for #8824 and it would be nice to have them both in the following release.

**Review by Maffooch — APPROVED** [review]

**Comment by mtesauro** [comment]
@kiblik 
> Ok, I will change the implementation to what @Maffooch offered. But I will add an option to customize a list of events via EnvVars (loaded in Settings) with default `["user_mentioned", "review_requested"]` - to allow admins to easily change this behavior.

Fair enough - that's a great middle ground. Thanks for that addition :+1:

**Review by mtesauro — APPROVED** [review]
Approved

File: unittests/test_notifications.py [hunk 1]
```diff
@@ -55,3 +58,46 @@ def test_merge_notifications_list(self):
...
+        notif_system, _ = Notifications.objects.get_or_create(user=None, template=False)
+
+        last_count = 0
+        with self.subTest('user off, system off'):
+            notif_user.user_mentioned = ()  # no alert
+            notif_user.save()
+            notif_system.user_mentioned = ()  # no alert
+            notif_system.save()
+            create_notification(event="user_mentioned", title="user_mentioned", recipients=['admin'])
+            self.assertEqual(mock.call_count, last_count + 0)
```
Comment (cneill) [inline unittests/test_notifications.py:74]: The `+ 0` doesn't seem necessary here?

```suggestion
            self.assertEqual(mock.call_count, last_count)
```

File: docs/content/en/integrations/notifications.md [hunk 2]
```diff
@@ -115,3 +115,13 @@ To activate notifications to Microsoft Teams, you have to:
 - Configure an Incoming Webhook in a Teams channel and copy the URL of the webhook to the clipboard
 - Activate `Enable Microsoft Teams notifications` in the System Settings
 - Paste the URL of the Incoming Webhook into the field `Msteams url`
+
+## Specific overrides
+
+System notification settings (scope: system) describe the sending of notifications to superadmins. User notification settings (scope: personal) describe sending notifications to the specific user.
+
+However, there is a specific use-case when the user decides to disable notifications (to decrease noise) but the system setting is used to override this behavior. These overrides apply only to `user_mentioned` and `review_requested`.
```
Comment (cneill) [inline docs/content/en/integrations/notifications.md:123]:
```suggestion
However, there is a specific use-case when the user decides to disable notifications (to decrease noise) but the system setting is used to override this behavior. These overrides apply only to `user_mentioned` and `review_requested` by default.
```

**Review by cneill — CHANGES_REQUESTED** [review]
Just a few small things, otherwise looks good

File: unittests/test_notifications.py (hunk 1 above)
Comment (kiblik) [reply unittests/test_notifications.py:74]: From the mathematical perspective, there is no difference. It just increases readability in the context of this unit test  - all other tests use `+ number` notation.
But if it is the only issue I can remove it.

**Review by cneill — APPROVED** [review]

**Comment by kiblik** [comment]
@mtesauro, can I ask for a merge please?

**Comment by mtesauro** [comment]
@kiblik Sure thing - merging now

**Pull request merged**